File: packages/emitter/testcases/real-world/async-ops/async-ops.ts
export async function delay(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

export async function fetchData(id: number): Promise<string> {
  await delay(100);
  return `Data for ID ${id}`;
}

export async function fetchMultiple(ids: number[]): Promise<string[]> {
  const promises = ids.map((id) => fetchData(id));
  return Promise.all(promises);
}

export async function processWithRetry<T>(
  fn: () => Promise<T>,
  maxRetries: number = 3
): Promise<T> {
  let lastError: Error | undefined;

  for (let i = 0; i < maxRetries; i++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error as Error;
      await delay(100 * (i + 1)); // Exponential backoff
    }
  }

  throw lastError || new Error("Max retries exceeded");
}

export class AsyncCache<K, V> {
  private cache: Map<K, V> = new Map();
  private loading: Map<K, Promise<V>> = new Map();

  async get(key: K, loader: () => Promise<V>): Promise<V> {
    // Check cache
    if (this.cache.has(key)) {
      return this.cache.get(key)!;
    }

    // Check if already loading
    if (this.loading.has(key)) {
      return this.loading.get(key)!;
    }

    // Start loading
    const promise = loader();
    this.loading.set(key, promise);

    try {
      const value = await promise;
      this.cache.set(key, value);
      return value;
    } finally {
      this.loading.delete(key);
    }
  }

  clear(): void {
    this.cache.clear();
    this.loading.clear();
  }
}

File: packages/emitter/testcases/real-world/async-ops/async-ops.cs
namespace TestCases.realworld.asyncops
{
    public class AsyncCache<K, V>
    {
        private global::Tsonic.JSRuntime.Map<K, V> cache = new global::Tsonic.JSRuntime.Map();

        private global::Tsonic.JSRuntime.Map<K, global::System.Threading.Tasks.Task<V>> loading = new global::Tsonic.JSRuntime.Map();

        public async global::System.Threading.Tasks.Task<V> get(K key, global::System.Func<global::System.Threading.Tasks.Task<V>> loader)
            {
            if (this.cache.has(key))
                {
                return this.cache.get(key)!;
                }
            if (this.loading.has(key))
                {
                return this.loading.get(key)!;
                }
            var promise = loader();
            this.loading.set(key, promise);
            try
            {
            var value = await promise;
            this.cache.set(key, value);
            return value;
            }
            finally
            {
            this.loading.delete(key);
            }
            }

        public void clear()
            {
            this.cache.clear();
            this.loading.clear();
            }
    }

            public static class asyncops
            {
                public static async global::System.Threading.Tasks.Task delay(double ms)
                    {
                    return new Promise((global::System.Action resolve) =>
                    {
                    setTimeout(resolve, ms);
                    });
                    }

                public static async global::System.Threading.Tasks.Task<string> fetchData(double id)
                    {
                    await delay(100.0);
                    return $"Data for ID {id}";
                    }

                public static async global::System.Threading.Tasks.Task<global::System.Collections.Generic.List<string>> fetchMultiple(global::System.Collections.Generic.List<double> ids)
                    {
                    var promises = global::Tsonic.JSRuntime.Array.map(ids, (double id) => fetchData(id));
                    return Promise.all(promises);
                    }

                public static async global::System.Threading.Tasks.Task<T> processWithRetry<T>(global::System.Func<global::System.Threading.Tasks.Task<T>> fn, double maxRetries = 3.0)
                    {
                    global::System.Exception? lastError;
                    for (int i = 0; i < maxRetries; i++)
                        {
                        try
                        {
                        return await fn();
                        }
                        catch (Exception error)
                        {
                        lastError = error;
                        await delay(100.0 * i + 1.0);
                        }
                        }
                    throw lastError ?? new Error("Max retries exceeded");
                    }
            }
}